File: system/src/lib/achievementEngine.js
/** 달성률 산정 엔진 — Linear / Flat / Custom */

export const ACHIEVEMENT_MODES = {
  LINEAR: 'linear',
  FLAT: 'flat',
  CUSTOM: 'custom',
}

export const GOAL_DIRECTIONS = {
  INCREASE: 'increase',
  DECREASE: 'decrease',
}

export const MODE_LABELS = {
  linear: 'Linear',
  flat: 'Flat',
  custom: 'Custom',
}

export const DIRECTION_LABELS = {
  increase: '증가',
  decrease: '감소',
}

/** Custom 모드용 달성률 산출식 프리셋 */
export const ACHIEVEMENT_PRESETS = [
  {
    id: 'standard_increase',
    label: '표준 달성률 (증가)',
    expr: '100 + (actual - target) / target * 100',
    direction: 'increase',
    hint: '100% + (실적-목표)/목표 × 100',
  },
  {
    id: 'standard_decrease',
    label: '표준 달성률 (감소)',
    expr: '100 + (target - actual) / target * 100',
    direction: 'decrease',
    hint: '100% + (목표-실적)/목표 × 100',
  },
  {
    id: 'ratio',
    label: '목표 대비 비율',
    expr: 'actual / target * 100',
    direction: 'increase',
    hint: '실적 ÷ 목표 × 100',
  },
  {
    id: 'capped_increase',
    label: '상한 120% (증가)',
    expr: 'min(120, 100 + (actual - target) / target * 100)',
    direction: 'increase',
    hint: '표준 달성률, 최대 120%',
  },
  {
    id: 'floor_decrease',
    label: '하한 80% (감소)',
    expr: 'max(80, 100 + (target - actual) / target * 100)',
    direction: 'decrease',
    hint: '표준 달성률, 최소 80%',
  },
]

export function round(value, digits = 2) {
  if (!Number.isFinite(value)) return null
  const unit = 10 ** digits
  return Math.round(value * unit) / unit
}

function parsePercentLike(value, fallback = null) {
  if (value == null || value === '') return fallback
  if (typeof value === 'number') return Number.isFinite(value) ? value : fallback
  const cleaned = String(value).replace(/[%\s,]/g, '').trim()
  if (!cleaned) return fallback
  const n = Number(cleaned)
  return Number.isFinite(n) ? n : fallback
}

export function applyAchievementPolicy(def, rawAchievement) {
  const raw = Number(rawAchievement)
  if (!Number.isFinite(raw)) return null

  const lowerCap = parsePercentLike(def?.capMin ?? def?.cap_min, 40)
  const upperCap = parsePercentLike(def?.capMax ?? def?.cap_max, 150)
  const scoreMultiplier = parsePercentLike(def?.scoreRule ?? def?.score_rule, 1)
  const adjustBand = parsePercentLike(def?.adjBand ?? def?.adj_band, 120)
  const adjustMultiplier = parsePercentLike(def?.penaltyRule ?? def?.penalty_rule, 0.1)

  // 승수는 100을 기준으로 초과/미달 폭에만 적용한다.
  let adjusted = 100 + ((raw - 100) * scoreMultiplier)

  if (adjusted < lowerCap) adjusted = lowerCap

  // 조정구간 이상이면 초과분에만 조정승수를 적용한다. 예: 130, band 120, x0.1 -> 121
  if (adjustBand != null && adjustMultiplier != null && adjusted >= adjustBand) {
    adjusted = adjustBand + ((adjusted - adjustBand) * adjustMultiplier)
  }

  if (adjusted > upperCap) adjusted = upperCap
  if (adjusted < lowerCap) adjusted = lowerCap
  return round(adjusted, 2)
}

export function normalizeGoalDirection(def) {
  if (def?.goalDirection === GOAL_DIRECTIONS.DECREASE || def?.goalDirection === 'decrease') {
    return GOAL_DIRECTIONS.DECREASE
  }
  if (def?.lowerIsBetter) return GOAL_DIRECTIONS.DECREASE
  return GOAL_DIRECTIONS.INCREASE
}

export function normalizeAchievementMode(def) {
  const mode = def?.achievementMode
  if (mode === ACHIEVEMENT_MODES.FLAT || mode === ACHIEVEMENT_MODES.CUSTOM) return mode
  return ACHIEVEMENT_MODES.LINEAR
}

export function isLeapYear(year) {
  const y = Number(year)
  return (y % 4 === 0 && y % 100 !== 0) || y % 400 === 0
}

export function daysInYear(year) {
  return isLeapYear(year) ? 366 : 365
}

/** 연초(1/1) ~ 해당 월 말일까지의 일수 */
export function daysToMonthEnd(year, month) {
  const y = Number(year)
  const m = Number(month)
  if (!Number.isFinite(y) || !Number.isFinite(m) || m < 1 || m > 12) return null
  const end = new Date(y, m, 0) // month-end
  const start = new Date(y, 0, 1)
  return Math.round((end - start) / 86400000) + 1
}

export function daysInMonth(year, month) {
  const y = Number(year)
  const m = Number(month)
  if (!Number.isFinite(y) || !Number.isFinite(m) || m < 1 || m > 12) return null
  return new Date(y, m, 0).getDate()
}

/** 평가시작월 ~ 해당 월 말일까지의 일수 */
export function daysFromPeriodStart(year, month, startMonth = 1) {
  const y = Number(year)
  const m = Number(month)
  const start = Number(startMonth) || 1
  if (!Number.isFinite(y) || !Number.isFinite(m) || m < start || m > 12) return null
  let days = 0
  for (let mo = start; mo <= m; mo += 1) {
    const d = daysInMonth(y, mo)
    if (d == null) return null
    days += d
  }
  return days
}

/** 평가시작월 1일 ~ 평가종료월 말일까지의 일수 */
export function daysInEvalPeriod(year, startMonth = 1, endMonth = 12) {
  const y = Number(year)
  const start = Number(startMonth) || 1
  const end = Number(endMonth) || 12
  if (!Number.isFinite(y) || start < 1 || start > 12 || end < 1 || end > 12 || start > end) return null
  let days = 0
  for (let mo = start; mo <= end; mo += 1) {
    const d = daysInMonth(y, mo)
    if (d == null) return null
    days += d
  }
  return days
}

export function normalizeEvalPeriod(def = {}) {
  let start = Number(def?.targetStartMonth ?? def?.target_start_month ?? 1)
  let end = Number(def?.targetEndMonth ?? def?.target_end_month ?? 12)
  if (!Number.isFinite(start) || start < 1 || start > 12) start = 1
  if (!Number.isFinite(end) || end < 1 || end > 12) end = 12
  if (start > end) [start, end] = [end, start]
  return { start, end }
}

export function isMonthInEvalPeriod(month, def = {}) {
  const m = Number(month)
  if (!Number.isFinite(m) || m < 1 || m > 12) return false
  const { start, end } = normalizeEvalPeriod(def)
  return m >= start && m <= end
}

/** 미리보기/저장용 — 의미 있는 월간목표 override만 주입 (기본 0은 Linear 계산을 막지 않음) */
export function resolveMonthlyTargetOverride(draft, selectedMonth) {
  const mt = draft?.monthlyTarget
  if (mt != null && mt !== '' && Number.isFinite(Number(mt))) {
    return { [selectedMonth]: Number(mt) }
  }
  return draft?.customMonthlyTargets || undefined
}

/** 월간 목표 산출 */
export function computeMonthlyTarget(def, month, year = def?.year) {
  const m = Number(month)
  if (!Number.isFinite(m) || m < 1 || m > 12) return null

  const { start, end } = normalizeEvalPeriod(def)
  if (m < start || m > end) return null

  const mode = normalizeAchievementMode(def)
  const annual = Number(def?.annualTarget ?? 0)
  const baseline = Number(def?.baselineActual ?? 0)
  const digits = annual < 10 ? 2 : 1

  // DB에 저장된 월간목표가 있으면 우선 사용
  const custom = def?.customMonthlyTargets?.[m] ?? def?.customMonthlyTargets?.[String(m)]
  if (custom != null && custom !== '') return round(Number(custom), digits)
  const legacy = def?.monthlyTargets?.[m] ?? def?.monthlyTargets?.[String(m)]
  if (legacy != null && legacy !== '') return round(Number(legacy), digits)

  if (mode === ACHIEVEMENT_MODES.FLAT) {
    return round(annual, digits)
  }

  if (mode === ACHIEVEMENT_MODES.CUSTOM) {
    return round(annual, digits)
  }

  // Linear: 기준실적 + (연간-기준) / 평가기간일수 × 경과일수
  const y = Number(year) || new Date().getFullYear()
  const elapsed = daysFromPeriodStart(y, m, start)
  const periodDays = daysInEvalPeriod(y, start, end)
  if (elapsed == null || !periodDays) return null
  return round(baseline + (annual - baseline) * (elapsed / periodDays), digits)
}

/** 1~12월 월간 목표 미리보기 */
export function buildMonthlyTargetPreview(def, year = def?.year) {
  return Array.from({ length: 12 }, (_, idx) => {
    const month = idx + 1
    return { month, target: computeMonthlyTarget(def, month, year) }
  })
}

/** 달성률 산출 */
export function calculateAchievementRate(def, actual, month, year = def?.year, filterValues = null) {
  const actualNum = Number(actual)
  if (!Number.isFinite(actualNum)) return null

  const target = computeMonthlyTarget(def, month, year)
  if (target == null) return null
  if (target === 0) return actualNum === 0 ? 100 : null

  const mode = normalizeAchievementMode(def)
  const direction = normalizeGoalDirection(def)
  let rawAchievement = null

  if (mode === ACHIEVEMENT_MODES.CUSTOM && def?.customAchievementExpr?.trim()) {
    const vars = {
      actual: actualNum,
      target,
      month: Number(month),
      annualTarget: Number(def?.annualTarget ?? 0),
      baseline: Number(def?.baselineActual ?? 0),
      yearProgress: (() => {
        const { start, end } = normalizeEvalPeriod(def)
        const span = end - start + 1
        return span > 0 ? (Number(month) - start + 1) / span : Number(month) / 12
      })(),
    }
    if (filterValues && typeof filterValues === 'object') {
      Object.assign(vars, filterValues)
    } else if (def?.filters && typeof def.filters === 'object') {
      for (let i = 1; i <= 30; i += 1) {
        const v = def.filters[i] ?? def.filters[String(i)]
        vars[`filter${i}`] = Number(v) || 0
      }
    }
    rawAchievement = evaluateAchievementExpression(def.customAchievementExpr, vars)
    return applyAchievementPolicy(def, rawAchievement)
  }

  if (direction === GOAL_DIRECTIONS.DECREASE) {
    rawAchievement = round(100 + ((target - actualNum) / target) * 100, 2)
    return applyAchievementPolicy(def, rawAchievement)
  }
  rawAchievement = round(100 + ((actualNum - target) / target) * 100, 2)
  return applyAchievementPolicy(def, rawAchievement)
}

/** Custom 산출식 — 허용 변수/함수만 사용 */
export function evaluateAchievementExpression(expr, vars = {}) {
  const raw = String(expr ?? '').trim()
  if (!raw) return null
  if (!/^[0-9+\-*/().%,\s_a-zA-Z]+$/.test(raw)) return null

  const safeVars = {
    actual: Number(vars.actual ?? 0),
    target: Number(vars.target ?? 0),
    month: Number(vars.month ?? 0),
    annualTarget: Number(vars.annualTarget ?? 0),
    baseline: Number(vars.baseline ?? 0),
    yearProgress: Number(vars.yearProgress ?? 0),
  }
  for (let i = 1; i <= 30; i += 1) {
    safeVars[`filter${i}`] = Number(vars[`filter${i}`] ?? 0)
  }

  try {
    const filterArgs = Array.from({ length: 30 }, (_, i) => `filter${i + 1}`)
    const fn = new Function(
      'actual', 'target', 'month', 'annualTarget', 'baseline', 'yearProgress',
      ...filterArgs,
      'min', 'max', 'abs', 'round',
      `"use strict"; return (${raw});`,
    )
    const result = fn(
      safeVars.actual,
      safeVars.target,
      safeVars.month,
      safeVars.annualTarget,
      safeVars.baseline,
      safeVars.yearProgress,
      ...filterArgs.map(k => safeVars[k]),
      Math.min,
      Math.max,
      Math.abs,
      (v, d = 2) => round(v, d),
    )
    return Number.isFinite(result) ? round(result, 2) : null
  } catch {
    return null
  }
}

/** eval config 기본값 추론 (기존 데이터 마이그레이션) */
export function inferEvalCalcDefaults(def = {}) {
  const monthly = def.monthlyTargets || {}
  const values = Array.from({ length: 12 }, (_, i) => {
    const v = monthly[i + 1] ?? monthly[String(i + 1)]
    return v == null ? null : Number(v)
  }).filter(v => v != null)

  let achievementMode = ACHIEVEMENT_MODES.LINEAR
  if (values.length >= 2) {
    const first = values[0]
    const allSame = values.every(v => Math.abs(v - first) < 0.0001)
    if (allSame) achievementMode = ACHIEVEMENT_MODES.FLAT
  }

  const goalDirection = def.lowerIsBetter || def.goalDirection === 'decrease'
    ? GOAL_DIRECTIONS.DECREASE
    : GOAL_DIRECTIONS.INCREASE

  const month1 = monthly[1] ?? monthly['1']
  const baselineActual = def.baselineActual != null
    ? Number(def.baselineActual)
    : (month1 != null ? Number(month1) : 0)

  const { start, end } = normalizeEvalPeriod(def)
  return {
    achievementMode,
    goalDirection,
    baselineActual,
    customAchievementExpr: def.customAchievementExpr || '',
    customMonthlyTargets: def.customMonthlyTargets || null,
    customTargetMode: def.customTargetMode || 'auto',
    targetStartMonth: start,
    targetEndMonth: end,
  }
}

export function enrichEvalConfigEntry(cfg = {}) {
  const inferred = inferEvalCalcDefaults(cfg)
  return {
    ...cfg,
    achievementMode: cfg.achievementMode || inferred.achievementMode,
    goalDirection: cfg.goalDirection || inferred.goalDirection,
    baselineActual: cfg.baselineActual != null ? Number(cfg.baselineActual) : inferred.baselineActual,
    customAchievementExpr: cfg.customAchievementExpr ?? inferred.customAchievementExpr,
    customMonthlyTargets: cfg.customMonthlyTargets ?? inferred.customMonthlyTargets,
    customTargetMode: cfg.customTargetMode || inferred.customTargetMode,
    targetStartMonth: cfg.targetStartMonth ?? cfg.target_start_month ?? inferred.targetStartMonth,
    targetEndMonth: cfg.targetEndMonth ?? cfg.target_end_month ?? inferred.targetEndMonth,
  }
}

File: server/achievement_engine.py
# -*- coding: utf-8 -*-
"""달성률·월목표 산정 엔진 (프론트 achievementEngine.js 와 동일 규칙)."""
from __future__ import annotations

import calendar
import json
import math
import re
from datetime import date
from typing import Any


ACHIEVEMENT_MODES = ("linear", "flat", "custom")
GOAL_DIRECTIONS = ("increase", "decrease")


def round_value(value: float | None, digits: int = 2) -> float | None:
    if value is None or not math.isfinite(value):
        return None
    unit = 10 ** digits
    return round(value * unit) / unit


def _parse_percent_like(value: Any, fallback: float) -> float:
    if value is None or value == "":
        return fallback
    try:
        parsed = float(str(value).replace("%", "").replace(",", "").strip())
        return parsed if math.isfinite(parsed) else fallback
    except (TypeError, ValueError):
        return fallback


def apply_achievement_policy(def_row: dict, raw_achievement: float | None) -> float | None:
    """상·하한, 승수, 조정구간을 프론트와 동일하게 적용한다."""
    if raw_achievement is None:
        return None
    try:
        raw = float(raw_achievement)
    except (TypeError, ValueError):
        return None
    if not math.isfinite(raw):
        return None

    lower = _parse_percent_like(def_row.get("cap_min"), 40.0)
    upper = _parse_percent_like(def_row.get("cap_max"), 150.0)
    multiplier = _parse_percent_like(def_row.get("score_rule"), 1.0)
    band = _parse_percent_like(def_row.get("adj_band"), 120.0)
    adjust = _parse_percent_like(def_row.get("penalty_rule"), 0.1)

    converted = 100.0 + ((raw - 100.0) * multiplier)
    converted = max(converted, lower)
    if converted >= band:
        converted = band + ((converted - band) * adjust)
    converted = min(max(converted, lower), upper)
    return round_value(converted, 2)


def is_leap_year(year: int) -> bool:
    return calendar.isleap(int(year))


def days_in_year(year: int) -> int:
    return 366 if is_leap_year(year) else 365


def days_to_month_end(year: int, month: int) -> int:
    y, m = int(year), int(month)
    if m < 1 or m > 12:
        raise ValueError("month must be 1..12")
    end = date(y, m, calendar.monthrange(y, m)[1])
    start = date(y, 1, 1)
    return (end - start).days + 1


def days_in_month(year: int, month: int) -> int:
    y, m = int(year), int(month)
    if m < 1 or m > 12:
        raise ValueError("month must be 1..12")
    return calendar.monthrange(y, m)[1]


def days_from_period_start(year: int, month: int, start_month: int = 1) -> int | None:
    y, m, start = int(year), int(month), int(start_month or 1)
    if m < start or m > 12:
        return None
    return sum(days_in_month(y, mo) for mo in range(start, m + 1))


def days_in_eval_period(year: int, start_month: int = 1, end_month: int = 12) -> int | None:
    y, start, end = int(year), int(start_month or 1), int(end_month or 12)
    if start < 1 or start > 12 or end < 1 or end > 12 or start > end:
        return None
    return sum(days_in_month(y, mo) for mo in range(start, end + 1))


def normalize_eval_period(def_row: dict) -> tuple[int, int]:
    start = int(def_row.get("target_start_month") or def_row.get("targetStartMonth") or 1)
    end = int(def_row.get("target_end_month") or def_row.get("targetEndMonth") or 12)
    if start < 1 or start > 12:
        start = 1
    if end < 1 or end > 12:
        end = 12
    if start > end:
        start, end = end, start
    return start, end


def normalize_mode(mode: str | None) -> str:
    m = str(mode or "linear").strip().lower()
    if m in ("flat", "custom"):
        return m
    return "linear"


def normalize_direction(direction: str | None, lower_is_better: bool = False) -> str:
    d = str(direction or "").strip().lower()
    if d == "decrease" or lower_is_better:
        return "decrease"
    return "increase"


def _parse_custom_monthly(raw: Any) -> dict[str, float]:
    if raw is None:
        return {}
    if isinstance(raw, dict):
        out = {}
        for k, v in raw.items():
            try:
                out[str(int(k))] = float(v)
            except Exception:
                continue
        return out
    if isinstance(raw, str) and raw.strip():
        try:
            return _parse_custom_monthly(json.loads(raw))
        except Exception:
            return {}
    return {}


def compute_monthly_target(def_row: dict, month: int, year: int | None = None) -> float | None:
    m = int(month)
    if m < 1 or m > 12:
        return None
    start, end = normalize_eval_period(def_row)
    if m < start or m > end:
        return None
    mode = normalize_mode(def_row.get("achievement_mode") or def_row.get("achievementMode"))
    annual = float(def_row.get("annual_target") if def_row.get("annual_target") is not None else def_row.get("annualTarget") or 0)
    baseline = float(def_row.get("baseline_actual") if def_row.get("baseline_actual") is not None else def_row.get("baselineActual") or 0)
    digits = 2 if abs(annual) < 10 else 1

    # DB에 저장된 월간목표가 있으면 그걸 우선 사용
    custom = _parse_custom_monthly(
        def_row.get("custom_monthly_targets_json")
        or def_row.get("customMonthlyTargets")
        or def_row.get("monthlyTargets")
    )
    if str(m) in custom:
        return round_value(custom[str(m)], digits)

    if mode == "flat":
        return round_value(annual, digits)

    if mode == "custom":
        return round_value(annual, digits)

    # Linear: baseline + (annual - baseline) / 평가기간일수 * 경과일수
    y = int(year if year is not None else def_row.get("year") or date.today().year)
    elapsed = days_from_period_start(y, m, start)
    period_days = days_in_eval_period(y, start, end)
    if elapsed is None or not period_days:
        return None
    return round_value(baseline + (annual - baseline) * elapsed / period_days, digits)


def build_monthly_targets_map(def_row: dict, year: int) -> dict[str, float]:
    """연간목표·산식 기준으로 1~12월 월간목표 맵을 생성한다."""
    out: dict[str, float] = {}
    # 생성 시 기존 custom 값을 무시하고 재산정하기 위해 복사본 사용
    base = {
        **def_row,
        "custom_monthly_targets_json": None,
        "customMonthlyTargets": None,
        "monthlyTargets": None,
    }
    for m in range(1, 13):
        v = compute_monthly_target(base, m, year)
        out[str(m)] = float(v if v is not None else 0)
    return out


def evaluate_expression(expr: str, vars_map: dict[str, float]) -> float | None:
    raw = str(expr or "").strip()
    if not raw:
        return None
    if not re.fullmatch(r"[0-9+\-*/().%,\s_a-zA-Z]+", raw):
        return None
    safe = {k: float(v) for k, v in vars_map.items() if isinstance(v, (int, float)) or str(v).replace(".", "", 1).lstrip("-").isdigit()}
    for i in range(1, 31):
        key = f"filter{i}"
        if key not in safe:
            safe[key] = 0.0
    for key in ("actual", "target", "month", "annualTarget", "baseline", "yearProgress"):
        if key not in safe:
            safe[key] = 0.0
    try:
        # Restrict names available to expression
        allowed = {
            **{k: safe[k] for k in safe},
            "min": min,
            "max": max,
            "abs": abs,
            "round": lambda v, d=2: round_value(float(v), int(d)),
        }
        result = eval(raw, {"__builtins__": {}}, allowed)  # noqa: S307 — controlled vars only
        return round_value(float(result), 2) if result is not None and math.isfinite(float(result)) else None
    except Exception:
        return None


def calculate_achievement_rate(
    def_row: dict,
    actual: float | None,
    month: int,
    year: int | None = None,
    filter_values: dict[str, float] | None = None,
) -> float | None:
    if actual is None:
        return None
    try:
        actual_num = float(actual)
    except Exception:
        return None
    if not math.isfinite(actual_num):
        return None

    target = compute_monthly_target(def_row, month, year)
    if target is None:
        return None
    if target == 0:
        return 100.0 if actual_num == 0 else None

    mode = normalize_mode(def_row.get("achievement_mode") or def_row.get("achievementMode"))
    direction = normalize_direction(
        def_row.get("goal_direction") or def_row.get("goalDirection"),
        bool(def_row.get("lowerIsBetter") or def_row.get("lower_is_better")),
    )
    y = int(year if year is not None else def_row.get("year") or date.today().year)

    if mode == "custom":
        expr = str(def_row.get("custom_achievement_expr") or def_row.get("customAchievementExpr") or "").strip()
        if expr:
            vars_map = {
                "actual": actual_num,
                "target": float(target),
                "month": float(month),
                "annualTarget": float(def_row.get("annual_target") or def_row.get("annualTarget") or 0),
                "baseline": float(def_row.get("baseline_actual") or def_row.get("baselineActual") or 0),
                "yearProgress": float(month) / 12.0,
            }
            if filter_values:
                vars_map.update(filter_values)
            return evaluate_expression(expr, vars_map)

    if direction == "decrease":
        return round_value(100 + ((target - actual_num) / target) * 100, 2)
    return round_value(100 + ((actual_num - target) / target) * 100, 2)


def evaluate_formula_expr(expr: str, operands: dict[str, float]) -> float | None:
    """가공식: operands(영문/한글 식별자)와 사칙연산."""
    raw = str(expr or "").strip()
    if not raw:
        return None
    if not re.fullmatch(r"[\w+\-*/().%\s]+", raw, re.UNICODE):
        return None
    safe = {
        str(k): float(v)
        for k, v in (operands or {}).items()
        if v is not None and math.isfinite(float(v))
    }
    try:
        result = eval(raw, {"__builtins__": {}}, {**safe, "min": min, "max": max, "abs": abs})  # noqa: S307
        return float(result) if result is not None and math.isfinite(float(result)) else None
    except Exception:
        return None
